core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -544,101 +550,204 @@ class EntityService(protected val ctx: RawlsRequestContext,
-      localProvider <- entityManager.resolveProviderFuture(EntityRequestArguments(workspaceContext, ctx))
+        // confirm if this is already a quicksilver workspace by checking settings
+        workspaceSettingService = workspaceSettingServiceConstructor.get.apply(ctx)
+        settings <- traceFutureWithParent("getWorkspaceSettings", s) { _ =>
+          workspaceSettingService.getWorkspaceSettings(workspaceName)
+        }
+        _ = if (
+          settings
+            .find(_.isInstanceOf[CompactDataTablesSetting])
+            .asInstanceOf[Option[CompactDataTablesSetting]]
+            .exists(_.config.enabled)
+        ) {
+          throw new RawlsExceptionWithErrorReport(ErrorReport("Quicksilver already enabled for this workspace"))
+        }
 
-      // get the list of entity types in this workspace
-      entityTypeMetadata <- localProvider.entityTypeMetadata(useCache = true, ctx)
+        // start a transaction; here's where we do a bunch of writes
+        userResult <- dataSource.inTransaction { dataAccess =>
+          val shardId: String = dataAccess.determineShard(workspaceId)
 
-      // start a transaction; here's where we do a bunch of writes
-      _ <- dataSource.inTransaction { dataAccess =>
-        val shardId: String = dataAccess.determineShard(workspaceContext.workspaceIdAsUUID)
+          val stopwatch = StopWatch.createStarted()
 
-        // loop over entity types
-        def allTypesResult: Iterable[ReadWriteAction[Iterator[Int]]] =
-          entityTypeMetadata.map { case (entityType, metadata) =>
-            logger.info(s"Quicksilver migration:     - $entityType (${metadata.count}) ...")
+          logger.info(
+            s"Quicksilver migration $workspaceId: starting (${stopwatch.formatTime()}) ..."
+          )
 
-            val thisTypeList: ReadAction[Seq[Entity]] = DBIO.from(
-              localProvider
-                .listEntities(entityType)
-                .runWith(Sink.seq)
-            )
+          for {
+            // get the current value of MySQL sort_buffer_size
+            defaultSortBufferSize <- dataAccess.compactEntityQuery.getSortBufferSetting
+            // batch into groups of $batchSize entities, currently 50k. This method returns the min and max entity ids
+            // for each batch. later queries will use those boundaries to migrate entities in batches, which
+            // prevents the temp tables from growing too large.
+            batchBoundaries <- quicksilverCalculateBatches(workspaceId, batchSize, dataAccess)
+            // niceties for logging
+            indexedBoundaries = batchBoundaries.zipWithIndex
+
+            // set sort buffer size to 8MB; this avoids MySQL errors during migration
+            // with a "Out of sort memory, consider increasing server sort buffer size" message.
+            // see also https://bugs.mysql.com/bug.php?id=103225
+            _ <- dataAccess.compactEntityQuery.setSessionSortBuffer(8388608L)
```

Comment:
This is my least favorite part of this PR overall. I think that the operation requiring sort memory is when we aggregate from the `QS_ATTR_TEMP` and insert into `QS_ENTITY_TEMP`, but I have not been able to eliminate the issue any way other than asking for more memory here.

We *could* increase `sort_buffer_size` globally and remove this code. Note though that this 8MB would be allocated for each db *connection*. Cloud console reports that Rawls is running with ~800 connections, so this would be about 10% of the db's total of 64G.